src/Microsoft.ML.FastTree/TreeEnsemble/RegressionTree.cs

```diff
@@ -14,25 +14,73 @@
 using Microsoft.ML.Model;
 using Microsoft.ML.Model.Pfa;
 using Newtonsoft.Json.Linq;
-using Float = System.Single;
 
 namespace Microsoft.ML.Trainers.FastTree.Internal
 {
-    public class RegressionTree
+    public class RegressionTreeView
     {
-        private double _maxOutput;
+        /// <summary>
+        /// <see cref="RegressionTreeView"/> is an immutable wrapper over <see cref="_tree"/> for exposing some tree's attribute to users.
+        /// </summary>
+        private readonly RegressionTree _tree;
+
+        // Immutable accessors to the underlying tree structure. They help users to inspect tree learned using, for example,
+        // gradient-boosting decision tree and random forest.
+        public ReadOnlySpan<int> LteChild => _tree.LteChild;
+        public ReadOnlySpan<int> GtChild => _tree.GtChild;
+        public ReadOnlySpan<int> NumericalSplitFeatureIndexes => _tree.SplitFeatures;
+        public ReadOnlySpan<float> NumericalSplitThresholds => _tree.RawThresholds;
+        public ReadOnlySpan<bool> CategoricalSplitFlags => _tree.CategoricalSplit;
+        public IReadOnlyList<IReadOnlyList<int>> CategoricalSplitFeatures => _tree.CategoricalSplitFeatures;
+        public IReadOnlyList<IReadOnlyList<int>> CategoricalSplitFeatureRanges => _tree.CategoricalSplitFeatureRanges;
```

Comment:
You're going to have to document this *really* carefully, and a proper implementation will be quite annoying. This sort of advanced structure is something that might be best achieved using an actual method, e.g., get features and ranges for this node, if applicable.

Reply:
Thanks for this nice suggestion. You will see
```csharp
        /// <summary>
        /// Return the training labels falling into the specified leaf.
        /// </summary>
        /// <param name="leafIndex">The index of the specified leaf.</param>
        /// <returns>Training labels</returns>
        public ReadOnlySpan<double> GetLeafSamplesAt(int leafIndex) => new ReadOnlySpan<double>(_leafSamples[leafIndex]);

        /// <summary>
        /// Return the weights for training labels falling into the specified leaf. If <see cref="GetLeafSamplesAt"/> and <see cref="GetLeafSampleWeightsAt"/>
        /// use the same input, the i-th returned value of this function is the weight of the i-th label in <see cref="GetLeafSamplesAt"/>.
        /// </summary>
        /// <param name="leafIndex">The index of the specified leaf.</param>
        /// <returns>Training labels' weights</returns>
        public ReadOnlySpan<double> GetLeafSampleWeightsAt(int leafIndex) => new ReadOnlySpan<double>(_leafSampleWeights[leafIndex]);
```
in the next iteration.

---
In reply to: [251182480](https://github.com/dotnet/machinelearning/pull/2243#discussion_r251182480) [](ancestors = 251182480)